### PR TITLE
Serve customers data via first API v1 endpoint in JSON:API format behind feature toggle

### LINKED ISSUE #8788: [API] Finalise Customer Endpoint
## Description

FR, UK and AU instances need to be able to create, find and update Customers via the API.

### Request Params
Nothing additional to the standard rq queries

### Response Attributes
Id - integer
Enterprise_id - integer
First_name - string
Last_name - string
Code - string
Email - string
Allow_charges - binary (NO CREATE OR UPDATE - force users to use the UI. False on create)
Tags - array of strings
Billing_address - address object. I believe we can use something already existing here but check in with product if needed.
Shipping_address - address object. I believe we can use something already existing here but check in with product if needed.
Terms_and_conditions_accepted_at - datetime (NO CREATE OR UPDATE - force users to use the UI. False on create.)

NOTE This field will not be added in the first iteration of this endpoint as there are significant performance considerations
Customer_balance - double (ON SHOW ONLY) (NO CREATE OR UPDATE)

### Scoping
This endpoint will be accessible to Enterprises only - authentication is granted to any enterprises the user owns or manages.
This endpoint will not have a scope for shoppers/non-enterprise users.
This endpoint will list all customers connected to enterprises that are owned or managed by the authenticated user

### Acceptance Criteria

1. Developer to provide a full list of RQ params that can be used. 
2. Create: Tester to confirm that customers created via the endpoint are correct, including all fields. Any fields that should not be creatable via API should be handled correctly.
3. Read: Tester to ensure RQ params filter correctly and Rs attributes are correct
4. Update: Tester to ensure that all fields update correctly. Any fields that should not be creatable via API should not change.
5. Delete: Tester to ensure deletes are handled correctly. Deletes should be singular via ID. No bulk deleting.
6. Tester to confirm scope. An enterprise user should see all and only of the customers related to enterprises they own or manage. No CRUD actions should be possible for customers relating enterprises they do not manage.
7. Product to confirm documentation is correct and sufficient. Improve if necessary.
8. Feature exists behind a feature toggle

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Closes #8788 

First release candidate of the customers endpoint. It's still behind a feature toggle so that we can test it in production before releasing it as stable.

I still have a few questions which I will add as comments to the code below. It would be great to clarify all these points in reviews.

#### What should we test?

- Visit `/api-docs` to see all supported requests and parameters.
- Visit `/admin/feature-toggle` to enable the `api_v1` feature.
- Check the acceptance criteria:
  > Developer to provide a full list of RQ params that can be used.
  > Create: Tester to confirm that customers created via the endpoint are correct, including all fields. Any fields that should not be creatable via API should be handled correctly.
  > Read: Tester to ensure RQ params filter correctly and Rs attributes are correct
  > Update: Tester to ensure that all fields update correctly. Any fields that should not be creatable via API should not change.
  > Delete: Tester to ensure deletes are handled correctly. Deletes should be singular via ID. No bulk deleting.
  > Tester to confirm scope. An enterprise user should see all and only of the customers related to enterprises they own or manage. No CRUD actions should be possible for customers relating enterprises they do not manage.
  > Product to confirm documentation is correct and sufficient. Improve if necessary.
  > Feature exists behind a feature toggle

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

#### Dependencies

Waiting for #https://github.com/openfoodfoundation/openfoodnetwork/issues/8787 which is included in this PR.

#### Documentation updates

Not yet. We have a new Swagger config which we can upload to Swagger Hub once this is public. We might even set CI up to publish that config automatically.

### CODE REVIEW

File: spec/requests/api/v1/customers_spec.rb [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        properties: CustomerSchema.writable_attributes,
+        required: CustomerSchema.required_attributes
+      }
+
+      response "201", "Minimal customer created" do
+        param(:customer) do
+          {
+            email: "test@example.com",
+            enterprise_id: enterprise1.id.to_s
+          }
```
Comment (mkllnk) [inline spec/requests/api/v1/customers_spec.rb:124]: I think that we are supposed to send parameters like this:
```suggestion
          {
            data: {
              type: "customer",
              attributes: {
                email: "test@example.com",
                enterprise_id: enterprise1.id.to_s
              }
            }
          }
```

> A resource can be created by sending a POST request to a URL that represents a collection of resources. The request MUST include a single [resource object](https://jsonapi.org/format/#document-resource-objects) as primary data. The [resource object](https://jsonapi.org/format/#document-resource-objects) MUST contain at least a type member.

https://jsonapi.org/format/#crud-creating

File: app/controllers/api/v1/base_controller.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        Bugsnag.notify(exception)
+
+        if Rails.env.development? || Rails.env.test?
+          render status: :unprocessable_entity,
+                 json: json_api_error(exception.message, meta: exception.backtrace)
+        else
+          render status: :unprocessable_entity,
+                 json: json_api_error(I18n.t(:unknown_error, scope: "api"))
+        end
+      end
```
Comment (mkllnk) [inline app/controllers/api/v1/base_controller.rb:59]: We respond with unprocessable entity when encountering a StandardError. I reckon that the StandardError case should only happen on unexpected errors in which case a 500 server error may be the right response here. Opinions?

**Review by mkllnk — COMMENTED** [review]
A couple of comments to discuss.

**Review by jibees — APPROVED** [review]
Looks good to me!

File: app/controllers/api/v1/base_controller.rb (hunk 2 above)
Comment (jibees) [reply app/controllers/api/v1/base_controller.rb]: I personally don't like the `unprocessable entity` error, which says nothing at all (and make me thing of an error due to client mistake).

**Comment by filipefurtad0** [comment]
Hey @mkllnk ,

tested:

- feature toggle:

With the feature toggled off, a valid request triggers a 404 error; toggling the feature on displays the expected result :+1: 

- GET[​/api​/v1​/customers] List customers - OK

- POST[​/api​/v1​/customers] Create customers

Looking at the acceptance criteria for this one _Tester to confirm that customers created via the endpoint are correct, including all fields. Any fields that should not be creatable via API should be handled correctly_ I'm not sure we're good here. Sending the request
`curl -X POST "https://staging.openfoodnetwork.org.uk/api/v1/customers?token=0_____________________________5" -H  "accept: application/json" -H  "Content-Type: application/json" -d "{\"enterprise_id\":1819,\"first_name\":\"Alice\",\"last_name\":\"Springs\",\"code\":\"BUYER1\",\"email\":\"alice@example.com\",\"tags\":[\"staff\",\"discount\"],\"billing_address\":null,\"shipping_address\":{\"phone\":\"0404 333 222 111\",\"latitude\":-37.8173751,\"longitude\":144.964803195704,\"first_name\":\"Alice\",\"last_name\":\"Springs\",\"street_address_1\":\"1 Flinders Street\",\"street_address_2\":\"\",\"postal_code\":\"1234\",\"locality\":\"Melbourne\",\"region\":\"Victoria\",\"country\":\"Australia\"}}"`

Leads to the error:

`{"errors":[{"detail":"Invalid resource. Please fix errors and try again."}],"meta":{"validation_errors":["Ship address state can't be blank"]}}`

- GET[​/api​/v1​/customers/{id}] Show customer - OK

Using the tokens from hubs which have a given user as customer allows retrieving the respective info. Requests with tokens from hubs which don't lead to the error `{"errors":[{"detail":"You are not authorized to perform that action."}]}` :+1: 

- PUT[​/api​/v1​/customers/{id}] Edit customer - OK, in general.

It seems that `shipping_address` fields need to be all edited at once, a similar validation error occurs, for all other fields which are not updated.

- DELETE[​/api​/v1​/customers/{id}] Delete customer - OK

- GET[​/api​/v1​/enterprises/{enterprise_id}/customers] List customers of an enterprise - OK

So, in summary this all looks good. The point I'm not so clear about is the create and update validation on `shipping_address`.

In any case, I think we're good to merge, as the feature toggle seems to work fine :+1:

**Comment by filipefurtad0** [comment]
As discussed just now with @jibees , it seems feature toggling the API selectively to some users is not working. Currently we can either fully enable or fully disable.

image

Is selective toggling a requirement for API/v1 and needs an issue? ping @openfoodfoundation/train-drivers-product-owners

**Comment by RachL** [comment]
I think I can workaround it for now if really no one is able to find their API key on their own.

**Comment by lin-d-hop** [comment]
Perhaps we just expose it and wait until we really need to feature toggle the API to implement any fixes needed

**Comment by RachL** [comment]
@lin-d-hop the endpoint is toggled, but it just cannot be toggled by user yet. It needs to be instance-wide. I think it's still good to keep it toggle for a while before exposing it? Unless we are ready already to support all instances in that process. In that case, forget my comment.

**Comment by mkllnk** [comment]
> it seems feature toggling the API selectively to some users is not working.

Ah, I just tested this and it works when you are logged in and use your authentication cookie. But it doesn't work via API key which is the normal use case...

This should be a very quick fix. #9098
